toffy/normalize.py

```diff
@@ -0,0 +1,238 @@
+                                 'file {}'.format(file_prefix, files[0], files[i]))
+
+    metrics = pd.concat(metrics)
+
+    metrics.to_csv(os.path.join(run_dir, file_prefix + '_combined.csv'), index=False)
+
+
+def combine_tuning_curve_metrics(dir_list):
+    """Combines metrics together into a single dataframe for fitting a turning curve
+
+    Args:
+        dir_list (list): list of directories to pull metrics from
+
+    Returns:
+        pd.DataFrame: dataframe containing aggregates metrics"""
+
+    # create list to hold all extracted data
+    all_dirs = []
+
+    # loop through directories, and if present, multiple fovs within directories
+    for dir in dir_list:
+
+        # generate aggregated table if it doesn't already exist
+        for prefix in ['pulse_heights', 'channel_counts']:
+            if not os.path.exists(os.path.join(dir, prefix + '_combined.csv')):
+                combine_run_metrics(dir, prefix)
+
+        # combine tables together
+        pulse_heights = pd.read_csv(os.path.join(dir, 'pulse_heights_combined.csv'))
+        channel_counts = pd.read_csv(os.path.join(dir, 'channel_counts_combined.csv'))
+        combined = pulse_heights.merge(channel_counts, 'outer', on=['fovs', 'masses'])
+
+        if len(combined) != len(pulse_heights):
+            raise ValueError("Pulse heights and channel counts must be generated for the same "
+                             "mass ranges and fovs. However, the data the following does do not "
+                             "exactly match: {}".format(dir))
+
+        # add directory label and add to list
+        combined['directory'] = dir
+        all_dirs.append(combined)
+
+    # combine data from each dir together
+    all_data = pd.concat(all_dirs)
+    all_data.reset_index()
+
+    # create normalized counts column
+    subset = all_data[['channel_counts', 'masses']]
+    all_data['norm_channel_counts'] = subset.groupby('masses').transform(lambda x: (x / x.max()))
+
+    return all_data
+
+
+def normalize_image_data(data_dir, output_dir, fovs, pulse_heights, panel_info,
+                         norm_func_path):
+    """Normalizes image data based on median pulse height from the run and a tuning curve
+
+    Args:
+        data_dir (str): directory with the image data
+        output_dir (str): directory where the normalized images will be saved
+        fovs (list or None): which fovs to include in normalization. If None, uses all fovs
+        pulse_heights: file containing pulse heights per mass per fov
+        panel_info_path: file containing mapping between channels and masses
+        norm_func_path: file containing the saved weights for the normalization function
+    """
+
+    # get FOVs to loop over
+    if fovs is None:
+        fovs = io_utils.list_folders(data_dir)
+
+    # load calibration function
+    with open(norm_func_path, 'r') as cf:
+        norm_json = json.load(cf)
```

Comment:
Might want to validate that `fovs` and `norm_func_path` are valid.

Reply:
Agreed. Remind me of this again on the next PR when we have the jupyter notebook included, then it will be obvious what should be validated where